File: rafiqwin/Rafiq.Core/Device.cs
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace Rafiq.Core;

/// <summary>
/// Everything that talks to the robot. It serves plain HTTP on the local
/// network, so nothing here should ever leave the house.
/// </summary>
public sealed class Device
{
    public Store Cfg { get; }
    readonly HttpClient _http;

    public Device(Store cfg)
    {
        Cfg = cfg;
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(6) };
    }

    public string Token { get => Store.GetToken(); set => Store.SetToken(value); }

    // what the robot last said about itself
    public bool? Reachable { get; private set; }
    public bool Paired, Linked, Following, Relaxing;
    public int FocusLeft, DndLeft;
    public string Version = "";
    public string Status = "";
    public event Action? Changed;

    DateTime _statusUntil = DateTime.MinValue;

    public void Flash(string s)
    {
        Status = s;
        _statusUntil = DateTime.UtcNow.AddSeconds(2.2);
        Changed?.Invoke();
    }
    public void TickStatus()
    {
        if (Status.Length > 0 && DateTime.UtcNow > _statusUntil) { Status = ""; Changed?.Invoke(); }
    }

    // ---- text ----

    public const int MaxLen = 84;

    /// <summary>
    /// The firmware keeps 84 bytes and cuts on a byte boundary, which would
    /// split a multi byte character in half. Trim by bytes here so it never
    /// has to. The panel draws ASCII, so anything fancier arrives intact but
    /// shows as blanks.
    /// </summary>
    public static string Clip(string s)
    {
        if (Encoding.UTF8.GetByteCount(s) <= MaxLen) return s;
        var sb = new StringBuilder();
        int n = 0;
        var e = System.Globalization.StringInfo.GetTextElementEnumerator(s);
        while (e.MoveNext())
        {
            var g = (string)e.Current;
            int c = Encoding.UTF8.GetByteCount(g);
            if (n + c > MaxLen - 3) break;      // three bytes for the ellipsis
            sb.Append(g); n += c;
        }
        return sb.ToString() + "…";
    }

    // ---- what the tiles do ----

    public Task Say(string raw)
    {
        var t = raw.Trim();
        if (t.Length == 0) return Task.CompletedTask;
        return Run("/api/msg", new() { ["m"] = Clip(t.Replace("\n", " ")) }, "Sent");
    }

    public Task Toast(string text, string kind, int seconds = 4) =>
        Run("/api/toast", new() { ["m"] = Clip(text), ["k"] = kind, ["s"] = seconds.ToString() }, null);

    public async Task StartFocus(int m)  { await Run("/api/focus", new() { ["m"] = m.ToString() }, $"Focus for {m} min"); FocusLeft = m * 60; }
    public async Task StopFocus()        { await Run("/api/focus", new() { ["m"] = "0" }, "Focus stopped"); FocusLeft = 0; }
    public async Task StartBreak(int m)  { await Run("/api/dnd", new() { ["m"] = m.ToString() }, $"On a break for {m} min"); DndLeft = m * 60; }
    public async Task EndBreak()         { await Run("/api/dnd", new() { ["m"] = "0" }, "Back"); DndLeft = 0; }
    public async Task SetRelax(bool on)  { await Run("/api/relax", new() { ["a"] = on ? "1" : "0" }, on ? "Resting" : "Back to normal"); Relaxing = on; }
    public async Task SetFollow(bool on) { await Run("/api/follow", new() { ["a"] = on ? "1" : "0" }, on ? "Watching the pointer" : "Eyes off"); Following = on; }
    public Task SetBusy(bool cam, bool mic) =>
        Run("/api/busy", new() { ["cam"] = cam ? "1" : "0", ["mic"] = mic ? "1" : "0" }, null);
    public Task Remind(string text) => Toast(text, "remind", 25);
    public Task CheckUpdate()       => Run("/api/update", new(), "Looking for an update");
    public async Task DeepSleep()   { await Run("/api/deepsleep", new(), "Going to sleep"); Linked = false; }

    public Task Canvas(byte[] px, int seconds = 8)
    {
        if (px.Length != 1024) { Flash("A screen is 1024 bytes"); return Task.CompletedTask; }
        return Run("/api/canvas",
                   new() { ["b"] = Convert.ToBase64String(px), ["s"] = seconds.ToString() }, "Drawn");
    }

    // ---- pairing ----

    public bool Pairing;
    public string PairError = "";

    /// <summary>
    /// Asking for a code needs no token, otherwise a robot whose token you
    /// had lost could never be paired again. Reading the code still means
    /// standing in front of it, which is the whole point.
    /// </summary>
    public async Task RequestCode()
    {
        if (Cfg.Ip.Length == 0) { Flash("Set the address first"); return; }
        PairError = "";
        try { await Post("/api/paircode", new()); Pairing = true; }
        catch { PairError = "Could not reach it"; }
        Changed?.Invoke();
    }

    public async Task Pair(string code)
    {
        var digits = new string(code.Where(char.IsDigit).ToArray());
        if (digits.Length != 6) { PairError = "Six digits"; Changed?.Invoke(); return; }
        try
        {
            var body = await Post("/api/pair", new() { ["c"] = digits });
            var t = Json.Str(body, "token");
            if (string.IsNullOrEmpty(t)) { PairError = "Wrong code"; Changed?.Invoke(); return; }
            Token = t; Paired = true; Pairing = false; PairError = "";
            Flash("Paired");
            await Refresh();
        }
        catch { PairError = "Wrong code, or it expired"; Changed?.Invoke(); }
    }

    public async Task Unpair()
    {
        await Run("/api/unpair", new(), "Unpaired");
        Token = ""; Paired = false;
    }

    // ---- state ----

    public async Task Refresh()
    {
        if (Cfg.Ip.Length == 0) { Reachable = null; Changed?.Invoke(); return; }
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, Url("/api/state"));
            req.Headers.TryAddWithoutValidation("X-Rafiq-App", "1");
            var tok = Token;
            if (tok.Length > 0) req.Headers.TryAddWithoutValidation("X-Rafiq-Token", tok);
            using var resp = await _http.SendAsync(req);
            int code = (int)resp.StatusCode;
            // 401 still means it answered, so the tray stays green and the
            // panel shows the pairing prompt rather than a dead robot.
            Reachable = code == 200 || code == 401;
            if (code == 401) { Paired = true; Linked = false; Changed?.Invoke(); return; }
            if (code != 200) { Changed?.Invoke(); return; }
            var s = await resp.Content.ReadAsStringAsync();
            Paired    = Json.Bool(s, "paired");
            Linked    = Json.Bool(s, "linked");
            Following = Json.Bool(s, "follow");
            Relaxing  = Json.Bool(s, "relax");
            FocusLeft = Json.Int(s, "focusLeft");
            DndLeft   = Json.Int(s, "dndLeft");
            var v = Json.Str(s, "fw");
            if (!string.IsNullOrEmpty(v)) Version = v;
        }
        catch { Reachable = false; Linked = false; }
        Changed?.Invoke();
    }

    // ---- transport ----

    async Task Run(string path, Dictionary<string, string> fields, string? say)
    {
        if (Cfg.Ip.Length == 0) { Flash("Set the address first"); return; }
        try
        {
            await Post(path, fields);
            Reachable = true;
            if (say != null) Flash(say);
        }
        catch (UnauthorizedAccessException) { Reachable = true; Flash("Pair with the robot first"); }
        catch { Reachable = false; Flash("Could not reach it"); }
        Changed?.Invoke();
    }

    Uri Url(string path) => new($"http://{Cfg.Ip.Trim()}{path}");

    async Task<string> Post(string path, Dictionary<string, string> fields)
    {
        using var req = new HttpRequestMessage(HttpMethod.Post, Url(path));
        req.Headers.TryAddWithoutValidation("X-Rafiq-App", "1");
        var tok = Token;
        if (tok.Length > 0) req.Headers.TryAddWithoutValidation("X-Rafiq-Token", tok);
        req.Content = new StringContent(Form(fields), Encoding.UTF8,
                                        "application/x-www-form-urlencoded");
        using var resp = await _http.SendAsync(req);
        int code = (int)resp.StatusCode;
        if (code is 401 or 403) throw new UnauthorizedAccessException();
        if (code != 200) throw new HttpRequestException($"status {code}");
        return await resp.Content.ReadAsStringAsync();
    }

    /// <summary>
    /// The device parses a plain form body, so everything outside the
    /// unreserved set has to be escaped, including the plus sign that a
    /// naive encoder would leave to be read back as a space.
    /// </summary>
    public static string Form(Dictionary<string, string> fields)
    {
        const string safe = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
        var sb = new StringBuilder();
        foreach (var (k, v) in fields)
        {
            if (sb.Length > 0) sb.Append('&');
            sb.Append(k).Append('=');
            foreach (var b in Encoding.UTF8.GetBytes(v))
            {
                char c = (char)b;
                if (safe.IndexOf(c) >= 0) sb.Append(c);
                else sb.Append('%').Append(b.ToString("X2"));
            }
        }
        return sb.ToString();
    }
}

/// <summary>
/// Small readers rather than a full parser: the device hand rolls its JSON
/// and one unexpected field should never cost us the rest of it.
/// </summary>
public static class Json
{
    public static string Str(string b, string key)
    {
        int k = b.IndexOf("\"" + key + "\"", StringComparison.Ordinal);
        if (k < 0) return "";
        int i = k + key.Length + 2;
        while (i < b.Length && (b[i] == ':' || b[i] == ' ')) i++;
        if (i >= b.Length || b[i] != '"') return "";
        i++;
        int e = b.IndexOf('"', i);
        return e < 0 ? "" : b.Substring(i, e - i);
    }
    public static string Raw(string b, string key)
    {
        int k = b.IndexOf("\"" + key + "\"", StringComparison.Ordinal);
        if (k < 0) return "";
        int i = k + key.Length + 2;
        while (i < b.Length && (b[i] == ':' || b[i] == ' ')) i++;
        int s = i;
        while (i < b.Length && b[i] != ',' && b[i] != '}') i++;
        return b.Substring(s, i - s).Trim();
    }
    public static bool Bool(string b, string key) => Raw(b, key) == "true";
    public static int  Int(string b, string key)  => int.TryParse(Raw(b, key), out var v) ? v : 0;
}

/// <summary>
/// Where the pointer is, sent as bare datagrams. Ten a second through a
/// fresh handshake each time would drown the robot, and a lost reading
/// costs nothing because another is a tenth of a second behind it.
/// </summary>
public sealed class PointerStream : IDisposable
{
    UdpClient? _udp;
    IPEndPoint? _to;

    public void Start(string host)
    {
        Stop();
        var bare = host.Split(':')[0].Trim();
        if (!IPAddress.TryParse(bare, out var ip))
        {
            try { ip = Dns.GetHostAddresses(bare).First(a => a.AddressFamily == AddressFamily.InterNetwork); }
            catch { return; }
        }
        _udp = new UdpClient();
        _to = new IPEndPoint(ip, 4210);
    }

    public void Send(int x, int y)
    {
        if (_udp == null || _to == null) return;
        var b = Encoding.ASCII.GetBytes($"{x} {y}");
        try { _udp.Send(b, b.Length, _to); } catch { }
    }

    public void Stop() { _udp?.Dispose(); _udp = null; _to = null; }
    public void Dispose() => Stop();
}

File: rafiqbar/Sources/Device.swift
import Foundation
import SwiftUI
import Network

/// Everything that talks to the robot. It serves plain HTTP on the local
/// network, which is why the bundle asks for local network access and allows
/// local loads only: nothing here should ever leave the house.
@MainActor
final class Device: ObservableObject {
    static let shared = Device()

    @AppStorage("deviceIP")    var ip: String = ""
    @AppStorage("watchClip")   var watchClipboard: Bool = false
    @AppStorage("breakMins")   var breakMins: Int = 45
    @AppStorage("breakOn")     var breakOn: Bool = false
    @AppStorage("focusMins")   var focusMins: Int = 25
    @AppStorage("breakCustom") var breakCustom: Int = 25
    @AppStorage("lockIdle")    var lockWhenIdle: Bool = false
    @AppStorage("lockIdleMin") var lockIdleMins: Int = 5
    @AppStorage("watchAV")     var watchAV: Bool = false
    @AppStorage("theme")       var theme: String = "system"
    @AppStorage("phrases")     var phrasesRaw: String =
        "On my way\nBack in 5\nIn a meeting\nCall me\nDone"

    var phrases: [String] {
        phrasesRaw.split(separator: "\n").map(String.init)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
    }

    var colorScheme: ColorScheme? {
        switch theme {
        case "light": return .light
        case "dark":  return .dark
        default:      return nil
        }
    }

    @Published var status: String = ""
    @Published var reachable: Bool? = nil
    @Published var busy = false

    /// What the robot last told us about itself. Everything the grid shows
    /// comes from here, so a tile is never lit unless the device agrees.
    @Published var paired = false
    @Published var linked = false
    @Published var following = false
    @Published var relaxing = false
    @Published var focusLeft = 0
    @Published var dndLeft = 0
    @Published var version = ""

    /// Set while a six digit code is on the robot's panel.
    @Published var pairing = false
    @Published var pairError = ""

    var token: String {
        get { Keychain.get("token") }
        set { Keychain.set(newValue, for: "token") }
    }

    // ---------------------------------------------------------------
    //  text
    // ---------------------------------------------------------------

    /// The firmware keeps 84 bytes and cuts on a byte boundary, which would
    /// split a multi byte character in half. Trim by bytes here so it never
    /// has to. The panel draws ASCII, so anything fancier arrives intact but
    /// shows as blanks.
    static let maxLen = 84
    static func clip(_ s: String) -> String {
        if s.utf8.count <= maxLen { return s }
        var out = ""
        var n = 0
        for ch in s {
            let c = String(ch).utf8.count
            if n + c > maxLen - 3 { break }        // 3 bytes for the ellipsis
            out.append(ch); n += c
        }
        return out + "\u{2026}"
    }

    // ---------------------------------------------------------------
    //  what the grid does
    // ---------------------------------------------------------------

    func say(_ raw: String) async {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        await run("/api/msg", ["m": Self.clip(text.replacingOccurrences(of: "\n", with: " "))],
                  say: "Sent")
    }

    /// A copy, a paste, or a word about standing up. None of it disturbs the
    /// message the robot is holding, which is yours.
    func toast(_ text: String, kind: String, seconds: Int = 4) async {
        await run("/api/toast",
                  ["m": Self.clip(text), "k": kind, "s": String(seconds)], say: nil)
    }

    func startFocus(_ mins: Int) async {
        await run("/api/focus", ["m": String(mins)], say: "Focus for \(mins) min")
        focusLeft = mins * 60
    }
    func stopFocus() async {
        await run("/api/focus", ["m": "0"], say: "Focus stopped")
        focusLeft = 0
    }

    func setRelax(_ on: Bool) async {
        await run("/api/relax", ["a": on ? "1" : "0"], say: on ? "Resting" : "Back to normal")
        relaxing = on
    }

    func setFollow(_ on: Bool) async {
        await run("/api/follow", ["a": on ? "1" : "0"], say: on ? "Watching the pointer" : "Eyes off")
        following = on
    }

    /// On a break: the robot holds the sign, the Mac locks itself.
    func startBreak(_ mins: Int) async {
        await run("/api/dnd", ["m": String(mins)], say: "On a break for \(mins) min")
        dndLeft = mins * 60
    }
    func endBreak() async {
        await run("/api/dnd", ["m": "0"], say: "Back")
        dndLeft = 0
    }

    /// One flag per device. Nothing about what is being said or seen.
    func setBusy(cam: Bool, mic: Bool) async {
        await run("/api/busy", ["cam": cam ? "1" : "0", "mic": mic ? "1" : "0"], say: nil)
    }

    func remind(_ text: String) async {
        await toast(text, kind: "remind", seconds: 25)
    }

    func checkUpdate() async {
        await run("/api/update", [:], say: "Looking for an update")
    }

    /// Nothing wakes it from this but the power switch, which is what was
    /// asked for, so the panel is what confirms it rather than this app.
    func deepSleep() async {
        await run("/api/deepsleep", [:], say: "Going to sleep")
        linked = false
    }

    /// One screenful of pixels, 128 by 64, a bit each, top row first.
    func canvas(_ bytes: Data, seconds: Int = 8) async {
        guard bytes.count == 1024 else { flash("A screen is 1024 bytes"); return }
        await run("/api/canvas", ["b": bytes.base64EncodedString(), "s": String(seconds)], say: "Drawn")
    }

    // ---------------------------------------------------------------
    //  pairing
    // ---------------------------------------------------------------

    /// Asking for a code needs no token, otherwise a device whose token you
    /// had lost could never be paired again. Reading the code still means
    /// standing in front of the thing, which is the whole point of it.
    func requestCode() async {
        guard !ip.isEmpty else { flash("Set the address first"); return }
        pairError = ""
        do {
            try await post("/api/paircode", [:])
            pairing = true
        } catch {
            pairError = "Could not reach it"
        }
    }

    func pair(with code: String) async {
        let digits = code.filter(\.isNumber)
        guard digits.count == 6 else { pairError = "Six digits"; return }
        do {
            let body = try await post("/api/pair", ["c": digits])
            guard let t = Self.jsonString(body, "token"), !t.isEmpty else {
                pairError = "Wrong code"; return
            }
            token = t
            paired = true
            pairing = false
            pairError = ""
            flash("Paired")
            await refresh()
        } catch {
            pairError = "Wrong code, or it expired"
        }
    }

    func unpair() async {
        await run("/api/unpair", [:], say: "Unpaired")
        token = ""
        paired = false
    }

    // ---------------------------------------------------------------
    //  state
    // ---------------------------------------------------------------

    func refresh() async {
        guard !ip.isEmpty else { reachable = nil; return }
        do {
            var req = URLRequest(url: try url("/api/state"))
            req.timeoutInterval = 3
            req.setValue("1", forHTTPHeaderField: "X-Rafiq-App")
            if !token.isEmpty { req.setValue(token, forHTTPHeaderField: "X-Rafiq-Token") }
            let (data, resp) = try await URLSession.shared.data(for: req)
            let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
            // 401 still means the robot answered, so the dot stays green and
            // the panel shows the pairing prompt instead of a dead device.
            reachable = (code == 200 || code == 401)
            guard code == 200, let s = String(data: data, encoding: .utf8) else {
                if code == 401 { paired = true; linked = false }
                return
            }
            paired    = Self.jsonBool(s, "paired")
            linked    = Self.jsonBool(s, "linked")
            following = Self.jsonBool(s, "follow")
            relaxing  = Self.jsonBool(s, "relax")
            focusLeft = Self.jsonInt(s, "focusLeft")
            dndLeft   = Self.jsonInt(s, "dndLeft")
            version   = Self.jsonString(s, "fw") ?? version
        } catch {
            reachable = false
            linked = false
        }
    }

    // ---------------------------------------------------------------
    //  transport
    // ---------------------------------------------------------------

    private var statusClear: Task<Void, Never>?

    private func run(_ path: String, _ fields: [String: String], say: String?) async {
        guard !ip.isEmpty else { flash("Set the address first"); return }
        busy = true
        defer { busy = false }
        do {
            _ = try await post(path, fields)
            reachable = true
            if let say { flash(say) }
        } catch let e as URLError where e.code == .userAuthenticationRequired {
            reachable = true
            flash("Pair with the robot first")
        } catch {
            reachable = false
            flash("Could not reach it")
        }
    }

    private func url(_ path: String) throws -> URL {
        let host = ip.trimmingCharacters(in: .whitespaces)
        guard let u = URL(string: "http://\(host)\(path)") else { throw URLError(.badURL) }
        return u
    }

    @discardableResult
    private func post(_ path: String, _ fields: [String: String]) async throws -> String {
        var req = URLRequest(url: try url(path))
        req.httpMethod = "POST"
        req.timeoutInterval = 6
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        req.setValue("1", forHTTPHeaderField: "X-Rafiq-App")
        if !token.isEmpty { req.setValue(token, forHTTPHeaderField: "X-Rafiq-Token") }
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        req.httpBody = fields
            .map { k, v in "\(k)=\(v.addingPercentEncoding(withAllowedCharacters: allowed) ?? "")" }
            .joined(separator: "&")
            .data(using: .utf8)
        let (data, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if code == 401 || code == 403 { throw URLError(.userAuthenticationRequired) }
        guard code == 200 else { throw URLError(.badServerResponse) }
        return String(data: data, encoding: .utf8) ?? ""
    }

    func flash(_ s: String) {
        status = s
        statusClear?.cancel()
        statusClear = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 2_200_000_000)
            guard !Task.isCancelled else { return }
            await MainActor.run { self?.status = "" }
        }
    }

    // ---------------------------------------------------------------
    //  Small readers rather than a decoder: the device hand rolls its
    //  JSON and one unexpected field should never cost us the rest.
    // ---------------------------------------------------------------

    static func jsonString(_ b: String, _ key: String) -> String? {
        guard let r = b.range(of: "\"\(key)\"") else { return nil }
        var i = r.upperBound
        while i < b.endIndex, b[i] == ":" || b[i] == " " { i = b.index(after: i) }
        guard i < b.endIndex, b[i] == "\"" else { return nil }
        i = b.index(after: i)
        guard let end = b[i...].firstIndex(of: "\"") else { return nil }
        return String(b[i..<end])
    }
    static func jsonRaw(_ b: String, _ key: String) -> String? {
        guard let r = b.range(of: "\"\(key)\"") else { return nil }
        var i = r.upperBound
        while i < b.endIndex, b[i] == ":" || b[i] == " " { i = b.index(after: i) }
        var out = ""
        while i < b.endIndex, b[i] != ",", b[i] != "}" { out.append(b[i]); i = b.index(after: i) }
        return out.trimmingCharacters(in: .whitespaces)
    }
    static func jsonBool(_ b: String, _ key: String) -> Bool { jsonRaw(b, key) == "true" }
    static func jsonInt(_ b: String, _ key: String) -> Int { Int(jsonRaw(b, key) ?? "") ?? 0 }
}

// ===================================================================
//  The pointer, sent over UDP
// ===================================================================
/// Ten a second through a fresh HTTP handshake each time would drown the
/// device, and a lost reading costs nothing because another is a tenth of a
/// second behind it. So these go out as bare datagrams and nothing is
/// acknowledged or retried.
@MainActor
final class Cursor {
    private var conn: NWConnection?
    private var timer: Timer?
    private var host = ""

    func start(host: String) {
        stop()
        self.host = host
        guard let p = NWEndpoint.Port(rawValue: 4210) else { return }
        let c = NWConnection(host: NWEndpoint.Host(host), port: p, using: .udp)
        c.start(queue: .global(qos: .utility))
        conn = c
        timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.tick() }
        }
    }

    func stop() {
        timer?.invalidate(); timer = nil
        conn?.cancel(); conn = nil
    }

    private func tick() {
        guard let conn else { return }
        let p = NSEvent.mouseLocation
        // The screen the pointer is actually on, so a second monitor does
        // not send the eyes hard over to one side and leave them there.
        let screen = NSScreen.screens.first { $0.frame.contains(p) } ?? NSScreen.main
        guard let f = screen?.frame, f.width > 0, f.height > 0 else { return }
        let nx = Float((p.x - f.minX) / f.width) * 2 - 1
        // AppKit measures up from the bottom and the robot looks down from
        // the top of the monitor, so this is flipped.
        let ny = 1 - Float((p.y - f.minY) / f.height) * 2
        let msg = "\(Int(nx * 1000)) \(Int(ny * 1000))"
        conn.send(content: msg.data(using: .utf8), completion: .idempotent)
    }
}
